google/cloud/forseti/services/scanner/dao.py

```diff
@@ -53,16 +55,17 @@ class Violation(base):
 
         __tablename__ = violations_tablename
 
+        full_name = Column(String(1024))
         id = Column(Integer, primary_key=True)
+        inventory_data = Column(Text(16777215))
         inventory_index_id = Column(String(256))
         resource_id = Column(String(256), nullable=False)
         resource_type = Column(String(256), nullable=False)
-        full_name = Column(String(1024))
         rule_name = Column(String(256))
         rule_index = Column(Integer, default=0)
-        violation_type = Column(String(256), nullable=False)
         violation_data = Column(Text)
-        inventory_data = Column(Text(16777215))
+        violation_type = Column(String(256), nullable=False)
+        violation_hash = Column(String(256))
```

Comment:
I think we need to make this column to be unique, if the hash already exists, the insert will fail.

Reply:
The question is, should we keep a record each time scanner finds a violation, even if it is the same one, and rely on downstream systems to filter out violations they've already seen? Or only store new violations?

Reply:
Great question. I think allowing downstream readers to build functionality around having multiple violations and filtering unwanted ones allows for innovation on this. I support not enforcing uniqueness on this field.

Reply:
We have decided that we will allow duplicate violations to be stored when scanners are run and find them, as this will allow downstream users the flexibility to de-dupe them or not, based on their use cases.